include/boost/multiprecision/cpp_int/misc.hpp

```diff
@@ -495,14 +495,491 @@ eval_gcd(
 {
    eval_gcd(result, a, static_cast<limb_type>(v < 0 ? -v : v));
 }
+//
+// What follows is Lehmer's GCD algorithm:
+// Essentially this uses the leading digit(s) of U and V
+// only to run a "simulated" Euclid algorithm.  It stops
+// when the calculated quotient differs from what would have been
+// the true quotient.  At that point the cosequences are used to
+// calculate the new U and V.  A nice lucid description appears
+// in "An Analysis of Lehmer's Euclidean GCD Algorithm",
+// by Jonathan Sorenson.  https://www.researchgate.net/publication/2424634_An_Analysis_of_Lehmer%27s_Euclidean_GCD_Algorithm
+// DOI: 10.1145/220346.220378.
+//
+// There are two versions of this algorithm here, and both are "double digit"
+// variations: which is to say if there are k bits per limb, then they extract
+// 2k bits into a double_limb_type and then run the algorithm on that.  The first
+// version is a straightforward version of the algorithm, and is designed for
+// situations where double_limb_type is a native integer (for example where
+// limb_type is a 32-bit integer on a 64-bit machine).  For 32-bit limbs it
+// reduces the size of U by about 30 bits per call.  The second is a more complex
+// version for situations where double_limb_type is a synthetic type: for example
+// __int128.  For 64 bit limbs it reduces the size of U by about 62 bits per call.
+//
+// The complexity of the algorithm given by Sorenson is roughly O(ln^2(N)) for
+// two N bit numbers.
```

Comment:
Wait, isn't the benchmark showing O(N^2) complexity for two N-bit numbers?

Reply:
It alternates from run to run between O(n^2) and O(NlnN), this is true of GMP also.  I don't think that either complexity is actually correct.